// @parcel/package-manager/lib/NodeResolverSync.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.NodeResolverSync = void 0;

function _path() {
  const data = _interopRequireDefault(require("path"));

  _path = function () {
    return data;
  };

  return data;
}

function _NodeResolverBase() {
  const data = require("./NodeResolverBase");

  _NodeResolverBase = function () {
    return data;
  };

  return data;
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class NodeResolverSync extends _NodeResolverBase().NodeResolverBase {
  resolve(id, from) {
    if (id[0] === '.') {
      id = _path().default.resolve(from, id);
    }

    let res;

    if (_path().default.isAbsolute(id)) {
      res = this.loadRelative(id);
    } else {
      res = this.loadNodeModules(id, from);
    }

    if (!res) {
      let e = new Error(`Could not resolve module "${id}" from "${from}"`); // $FlowFixMe

      e.code = 'MODULE_NOT_FOUND';
      throw e;
    }

    return res;
  }

  loadRelative(id) {
    // Find a package.json file in the current package.
    let pkg = this.findPackage(_path().default.dirname(id)); // First try as a file, then as a directory.

    return this.loadAsFile(id, pkg) || this.loadDirectory(id, pkg);
  }

  findPackage(dir) {
    // Find the nearest package.json file within the current node_modules folder
    let pkgFile = this.fs.findAncestorFile(['package.json'], dir);

    if (pkgFile != null) {
      return this.readPackage(pkgFile);
    }
  }

  readPackage(file) {
    let cached = this.packageCache.get(file);

    if (cached) {
      return cached;
    }

    let json = this.fs.readFileSync(file, 'utf8');
    let pkg = JSON.parse(json);
    this.packageCache.set(file, pkg);
    return pkg;
  }

  loadAsFile(file, pkg) {
    // Try all supported extensions
    let found = this.fs.findFirstFile(this.expandFile(file));

    if (found) {
      return {
        resolved: found,
        pkg
      };
    }

    return null;
  }

  loadDirectory(dir, pkg = null) {
    try {
      pkg = this.readPackage(dir + '/package.json'); // Get a list of possible package entry points.

      let entries = this.getPackageEntries(dir, pkg);

      for (let file of entries) {
        // First try loading package.main as a file, then try as a directory.
        const res = this.loadAsFile(file, pkg) || this.loadDirectory(file, pkg);

        if (res) {
          return res;
        }
      }
    } catch (err) {// ignore
    } // Fall back to an index file inside the directory.


    return this.loadAsFile(_path().default.join(dir, 'index'), pkg);
  }

  loadNodeModules(id, from) {
    try {
      let module = this.findNodeModulePath(id, from);

      if (!module || module.resolved) {
        return module;
      } // If a module was specified as a module sub-path (e.g. some-module/some/path),
      // it is likely a file. Try loading it as a file first.


      if (module.subPath) {
        let pkg = this.readPackage(module.moduleDir + '/package.json');
        let res = this.loadAsFile(module.filePath, pkg);

        if (res) {
          return res;
        }
      } // Otherwise, load as a directory.


      if (module.filePath) {
        return this.loadDirectory(module.filePath);
      }
    } catch (e) {// ignore
    }
  }

}

exports.NodeResolverSync = NodeResolverSync;
